src/models/StyledComponent.js

```diff
@@ -16,138 +16,186 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
-          { ...accum, [key]: typeof attrs[key] === 'function' ? attrs[key](context) : attrs[key] }
-        ), {})
-        return { ...context, ...this.attrs }
-      }
+      this.attrs = Object.keys(attrs).reduce((acc, key) => {
+        const attr = attrs[key]
+        // eslint-disable-next-line no-param-reassign
+        acc[key] = typeof attr === 'function' ? attr(context) : attr
+        return acc
+      }, {})
 
-      generateAndInjectStyles(theme: any, props: any) {
-        const executionContext = this.buildExecutionContext(theme, props)
-        return componentStyle.generateAndInjectStyles(executionContext)
-      }
+      return { ...context, ...this.attrs }
+    }
 
-      componentWillMount() {
-        // If there is a theme in the context, subscribe to the event emitter. This
-        // is necessary due to pure components blocking context updates, this circumvents
-        // that by updating when an event is emitted
-        if (this.context[CHANNEL]) {
-          const subscribe = this.context[CHANNEL]
-          this.unsubscribe = subscribe(nextTheme => {
-            // This will be called once immediately
-
-            // Props should take precedence over ThemeProvider, which should take precedence over
-            // defaultProps, but React automatically puts defaultProps on props.
-            const { defaultProps } = this.constructor
-            const isDefaultTheme = defaultProps && this.props.theme === defaultProps.theme
-            const theme = this.props.theme && !isDefaultTheme ? this.props.theme : nextTheme
-            const generatedClassName = this.generateAndInjectStyles(theme, this.props)
-            this.setState({ theme, generatedClassName })
-          })
-        } else {
-          const theme = this.props.theme || {}
-          const generatedClassName = this.generateAndInjectStyles(
-            theme,
-            this.props,
-          )
-          this.setState({ theme, generatedClassName })
-        }
+    generateAndInjectStyles(theme: any, props: any) {
+      const { componentStyle } = this.constructor
+      const executionContext = this.buildExecutionContext(theme, props)
+      const className = componentStyle.generateAndInjectStyles(executionContext)
+
+      if (typeof process !== 'undefined' && process.env.NODE_ENV !== 'production') {
+        const { warnTooManyClasses } = this.constructor
+        warnTooManyClasses(className)
       }
 
-      componentWillReceiveProps(nextProps: { theme?: Theme, [key: string]: any }) {
-        this.setState((oldState) => {
+      return className
+    }
+
+    componentWillMount() {
+      // If there is a theme in the context, subscribe to the event emitter. This
+      // is necessary due to pure components blocking context updates, this circumvents
+      // that by updating when an event is emitted
+      if (this.context[CHANNEL]) {
+        const subscribe = this.context[CHANNEL]
+        this.unsubscribe = subscribe(nextTheme => {
```

Comment:
Am I understanding correctly that if you use a theme, it subscribes to changes? Ouch. This can mean that *all* components are subscribed to theme changes, I'm going to guess that that is not great for resource use.

Since I never change themes in my app while running, I'd like this subscription to be optional.

I have a similar problem with translations, and I resorted to just re-mounting the entire app whenever the language changes (by using the language as a key on the root component). I don't want to pay the listener cost on every piece of displayed text.

How about providing a ThemeListener component that you can put below sCU components to update the tree below with the new context? Or adding a _listen prop on all components that decides whether to add this listener?